src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentGenericTypePass.cs

```diff
@@ -140,8 +140,11 @@ private void Process(ComponentIntermediateNode node)
             // not set the items. We won't be able to do type inference on this and so it will just be nonsense.
             foreach (var attribute in node.Attributes)
             {
-                if (attribute != null && TryFindGenericTypeNames(attribute.BoundAttribute, out var typeParameters))
+                if (attribute != null && TryFindGenericTypeNames(attribute.BoundAttribute, attribute.GloballyQualifiedTypeName, out var typeParameters))
                 {
+                    // Keep only type parameters defined by this component.
+                    typeParameters = typeParameters.Where(bindings.ContainsKey).ToArray();
```

Comment:
Can components be nested inside other types?

Reply:
I'm not sure, but I don't think so. At least I couldn't create such component and successfully reference it from `.razor`.

But anyway, I don't think it matters that much - this algorithm exists only to report error RZ10001 when type parameter inference isn't possible, but it doesn't have to be completely accurate nor is it today. (Not generating the razor error is completely fine as you would still get C# errors, the other way around just means you might need to specify the type parameters unnecessarily, one such issue is this PR fixing.)